### ISSUE TITLE
torch.std() returns nan for single item tensors.

### ISSUE THREAD

**Issue by erogol** [pytorch/pytorch#29372]
## 🐛 Bug

np.std(4) returns 0 whereas torch.std(torch.tensor(4)) returns NaN. This causes numerical instabilities in certain situations.

## To Reproduce

import numpy as np
np.std(4)  # returns 0

import torch
torch.std(torch.tensor(4.))  # returns NaN

## Expected behavior

It should also return 0 to be consistent.

## Environment

PyTorch 1.3

cc @ezyang @gchanan @zou3519 @jerryzh168

**Comment by SsnL** [comment]
Well, it is because the for np the default is `ddof=0`, but pytorch has default `unbiased=True`.

**Comment by SsnL** [comment]
This is more about numpy parity rather than correctness. Hence I removed the hipri flag.

**Comment by ezyang** [comment]
We should do this.

**Comment by vincentqb** [comment]
According to the [documentation](https://pytorch.org/docs/stable/torch.html#torch.std), the default is unbiased estimate (divide by `length-1`) for the standard deviation, so nan is expected.

```
>>> torch.std(torch.tensor(4.))
tensor(nan)
>>> torch.std(torch.tensor(4.), unbiased=False)
tensor(0.)
```

**Comment by piyush-kgp** [comment]
@vincentqb , then may be we should raise a division by zero error

**Comment by piyush-kgp** [comment]
Actually this got me curious as to what does `torch.tensor(0)/torch.tensor(0)` return. Surprisingly it didnt throw any error, rather my REPL crashed and closed the process in which python was running with this message on my terminal: `Floating point exception: 8`. 

Is this expected?

**Comment by ezyang** [comment]
That's because `torch.tensor(0)` infers to be an integer, and integer div-by-zero is UB. I think we have an issue filed somewhere to test this and fail more gracefully.

**Comment by piyush-kgp** [comment]
ok. I think you might be referring to this https://github.com/pytorch/pytorch/issues/327, which seems to be a work in progress

**Comment by heytanay** [comment]
I have tried solving this issue by adding a conditional in the function [_std(g, input, dim, unbiased, keepdim)](https://github.com/pytorch/pytorch/blob/3990e9d1ca23e39bfa65fe1d907ddb9f9dbf0919/torch/onnx/symbolic_opset9.py#L1810)  in `torch/onnx/symbolic_opset9.py`, but it doesn't seem to work.
Any guidance will be greatly appreciated, as I am a complete beginner and looking to forward to contribute here!

**Comment by albanD** [comment]
As @vincentqb mentioned above, we have a different definition of std than numpy.
If you set `ddof=1` in numpy to have the same behavior, you will get `nan` just like pytorch.

**Comment by vadimkantorov** [comment]
I propose to reconsider this, especially with the recent NumPy-compat push. I don't think that returning nan for reduction over unit-sized axes is useful for client code, returning 0 is more useful. So maybe special case unit-sized reduction?

```python
import torch

a = torch.tensor([[1, 2, 3, 4]], dtype = torch.float32)

print(a.std(0)) # [nan, nan, nan, nan]
print(a.numpy().std(0)) # [0, 0, 0, 0]
```

cc @mruberry
